Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -236,111 +515,238 @@ void MultipleScatteringCorrection::exec() {
+void MultipleScatteringCorrection::calculateL12s(const MultipleScatteringCorrectionDistGraber &distGraberContainer, //
+                                                 const MultipleScatteringCorrectionDistGraber &distGraberSample,    //
+                                                 std::vector<double> &L12sContainer,                                //
+                                                 std::vector<double> &L12sSample,                                   //
+                                                 const Geometry::IObject &shapeContainer,                           //
+                                                 const Geometry::IObject &shapeSample) {
+  const int64_t numVolumeElementsSample = distGraberSample.m_numVolumeElements;
+  const int64_t numVolumeElementsContainer = distGraberContainer.m_numVolumeElements;
+  const int64_t numVolumeElements = numVolumeElementsSample + numVolumeElementsContainer;
+  // L12 is a upper off-diagonal matrix from the hybrid of container and sample.
+  // e.g. container: a, b, c
+  //      sample: alpha, beta, gamma
+  //      hybrid: a, b, c, alpha, beta, gamma
+  //      L12:
+  //               a    b      c     alpha   beta    gamma
+  //      -------------------------------------------------
+  //      a     |  x   L12[0] L12[1] L12[2]  L12[3]  L12[4]
+  //      b     |  x    x     L12[5] L12[6]  L12[7]  L12[8]
+  //      c     |  x    x      x     L12[9]  L12[10] L12[11]
+  //      alpha |  x    x      x      x      L12[12] L12[13]
+  //      beta  |  x    x      x      x       x      L12[14]
+  //      gamma |  x    x      x      x       x       x
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputWS))
+  for (int64_t indexTo = 0; indexTo < numVolumeElements; ++indexTo) {
+    PARALLEL_START_INTERUPT_REGION
+    // find the position of first scattering element (container or sample)
+    const auto posTo = indexTo < numVolumeElementsContainer
+                           ? distGraberContainer.m_elementPositions[indexTo]
+                           : distGraberSample.m_elementPositions[indexTo - numVolumeElementsContainer];
+    // NOTE:
+    // We are using Track to ensure that the distance calculated are within the material
+    Track track(posTo, V3D{0, 0, 1}); // reusable track, eco-friendly
+    for (int64_t indexFrom = indexTo + 1; indexFrom < numVolumeElements; ++indexFrom) {
+      const int64_t idx = calcLinearIdxFromUpperTriangular(numVolumeElements, indexTo, indexFrom);
+      const auto posFrom = indexFrom < numVolumeElementsContainer
+                               ? distGraberContainer.m_elementPositions[indexFrom]
+                               : distGraberSample.m_elementPositions[indexFrom - numVolumeElementsContainer];
+      const V3D unitVector = getDirection(posFrom, posTo);
+
+      // combined
+      track.reset(posFrom, unitVector);
+      track.clearIntersectionResults();
+      const auto dist1_container = getDistanceInsideObject(shapeContainer, track);
```

Comment:
Same thing here about declaring it to be rays.

Thinking about it more we may want to have another version of `getDistanceInsideObject` that takes the two points, the geometry, and the track and does all of the magic together. It also might not be a terribly useful refactor.

Reply:
- the local vars are renamed as suggested.
- the missing reset between calls to getDistanceInsideObejct is added now.

If we decided to go with the partial caching for L12 (or drop caching for L12 completely), getting a new function that takes two element position and shape would be very helpful.
However, I am a little bit concerned with adding too many small helper functions as they tend to make the code harder and harder to maintaine due to code fragmentation.
I think we should carefully think through the new design before starting a refactoring process.

Reply:
I agree that balancing out size to maintain clarity is hard